[LV][NFC] Tidy up partial reduction tests with filter-out-after option

david-arm opened the pull request:
A few test files seemed to have been edited after using the
update_test_checks.py script, which can make life hard for
developers when trying to update these tests in future
patches. Also, the tests still had this comment at the top

; NOTE: Assertions have been autogenerated by ...

which could potentially be confusing, since they've not
strictly been auto-generated.

I've attempted to keep the spirit of the original tests by
excluding all CHECK lines after the scalar.ph IR block,
however I've done this by using a new option called
--filter-out-after to the update_test_checks.py script.

Mel-Chen reviewed (approved):
LGTM

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/partial-reduce-chained.ll:
```diff
@@ -1035,3 +1515,55 @@ attributes #0 = { vscale_range(1,16) }
 
 !0 = !{!"llvm.loop.vectorize.scalable.enable", i1 true}
 !1 = distinct !{!0}
+;.
```
Those are not really interesting. Could you pass —check-globals=none?

SamTebbs33 reviewed (commented):
I think the extra work to not stage these changes in git (`git add -p .` is great for this) is worth not testing code blocks that we don't actually care about.

david-arm commented:
> I think the extra work to not stage these changes in git (`git add -p .` is great for this) is worth not testing code blocks that we don't actually care about.

I guess it's just every time the IR changes innocuously due to some vplan transformation or whatever, the tests need to be regenerated and then the developer has to go back and delete the scalar loop for every test, and the more tests we add the more painful this becomes. I think if we don't want auto-generated tests then we should remove the `; NOTE` line from the top of the tests because they aren't autogenerated - they've been manually edited. @fhahn any thoughts?

If there is a simple command-line tool or script that can be run for maintaining these tests it's worth documenting that at the top of the test, since the "NOTE..." line suggests you just run the update_test_checks script.

fhahn reviewed (commented):
> > I think the extra work to not stage these changes in git (`git add -p .` is great for this) is worth not testing code blocks that we don't actually care about.
> 
> I guess it's just every time the IR changes innocuously due to some vplan transformation or whatever, the tests need to be regenerated and then the developer has to go back and delete the scalar loop for every test, and the more tests we add the more painful this becomes. I think if we don't want auto-generated tests then we should remove the `; NOTE` line from the top of the tests because they aren't autogenerated - they've been manually edited. @fhahn any thoughts?
> 
> If there is a simple command-line tool or script that can be run for maintaining these tests it's worth documenting that at the top of the test, since the "NOTE..." line suggests you just run the update_test_checks script.

If they are manually edited, we should drop the NOTE.  In general, IMO having more targeted checks instead of automatically checking everything makes it easier to for the reviewers, as there generally are less changes/noise in the diff, while it makes it more difficult to update. 

The trend overall in LLVM seems to be going more towards autogenerating everything and in the past we accepted a number of changes to go from manual check lines to auto-generated ones, especially when the manual check lines weren't comprehensive. 

It's not necessarily ideal in all cases for LV tests, as we end up checking a lot of 'uninteresting' things, and end up with much larger check files, so I think it is great when people try to go with more targeted checks. What may be helpful would be an extension to `update_test_checks.py` that allows filtering out some basic blocks; that way we could still auto-generate, but filter out the scalar loop.

david-arm commented:
> > > I think the extra work to not stage these changes in git (`git add -p .` is great for this) is worth not testing code blocks that we don't actually care about.
> > 
> > 
> > I guess it's just every time the IR changes innocuously due to some vplan transformation or whatever, the tests need to be regenerated and then the developer has to go back and delete the scalar loop for every test, and the more tests we add the more painful this becomes. I think if we don't want auto-generated tests then we should remove the `; NOTE` line from the top of the tests because they aren't autogenerated - they've been manually edited. @fhahn any thoughts?
> > If there is a simple command-line tool or script that can be run for maintaining these tests it's worth documenting that at the top of the test, since the "NOTE..." line suggests you just run the update_test_checks script.
> 
> If they are manually edited, we should drop the NOTE. In general, IMO having more targeted checks instead of automatically checking everything makes it easier to for the reviewers, as there generally are less changes/noise in the diff, while it makes it more difficult to update.
> 
> The trend overall in LLVM seems to be going more towards autogenerating everything and in the past we accepted a number of changes to go from manual check lines to auto-generated ones, especially when the manual check lines weren't comprehensive.
> 
> It's not necessarily ideal in all cases for LV tests, as we end up checking a lot of 'uninteresting' things, and end up with much larger check files, so I think it is great when people try to go with more targeted checks. What may be helpful would be an extension to `update_test_checks.py` that allows filtering out some basic blocks; that way we could still auto-generate, but filter out the scalar loop.

Yeah that's a good point. My main concern is the cost of maintaining tests have non-trivial IR that change due to a VPlan transformation, optimisation, difference in naming scheme, etc. It's imagine nobody wants to manually update 20 tests every time for something trivial! I'm also happy to look into the idea of changing update_test_checks.py to filter all CHECK lines from the start of a particular block.

david-arm commented:
I'll create a new PR soon to add new functionality to update_test_checks.py for stopping CHECK line generation from a particular line onwards! I have a downstream patch working.

fhahn commented on llvm/utils/UpdateTestChecks/common.py:
```diff
@@ -74,13 +74,17 @@ class Filter(Regex):
     """
 
-    def __init__(self, regex, is_filter_out):
+    def __init__(self, regex, is_filter_out, is_filter_all_after):
```
It's probably best to create a new, separate review for the change to the script, with a dedicated test for the script.

david-arm replied:
Yep, can do. I just thought I'd put the whole thing here first for context and to help reviewers understand the typical use case. I'll spin out the python changes in a new PR.

david-arm replied:
I've put up https://github.com/llvm/llvm-project/pull/129739

SamTebbs33 reviewed (approved):
This looks great to me, thanks!

david-arm committed:
[LV][NFC] Tidy up partial reduction tests with filter-out-after option

A few test files seemed to have been edited after using the
update_test_checks.py script, which can make life hard for
developers when trying to update these tests in future
patches. Also, the tests still had this comment at the top

; NOTE: Assertions have been autogenerated by ...

which could potentially be confusing, since they've not
strictly been auto-generated.

I've attempted to keep the spirit of the original tests by
excluding all CHECK lines after the scalar.ph IR block,
however I've done this by using a new option called
--filter-out-after to the update_test_checks.py script.

david-arm commented:
Rebased after landing https://github.com/llvm/llvm-project/pull/129739. This PR is now a test-only change!

SamTebbs33 reviewed (approved):
Thanks for doing that :)
